### DISCUSSION TITLE
cmake trying msvc -- why?  mingw given

### DISCUSSION THREAD

**Discussion by u-an-i** [albar965/littlenavmap#688] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/84718885/121771158-f5c2f300-cb6d-11eb-84f5-b4aec7ea955d.png)

cmake is accessing msvc. mingw is given.

what do i have to do to get marble made?

**Comment by albar965** [comment]
Sorry, I can only help with Mingw which I use to build on Linux and Windows. I don't have MSVC. Neither to I have much of a clue about cmake since I use qmake for all projects.

**Comment by u-an-i** [comment, marked as answer]
inside the **build target directory** on a cli as a 1-liner (so remove newlines):

**C:\Qt\Tools\CMake_64\bin\cmake**
  -DCMAKE_BUILD_TYPE=Release
  -DSTATIC_BUILD=TRUE
  -DQTONLY=TRUE
  -DBUILD_MARBLE_EXAMPLES=NO
  -DBUILD_INHIBIT_SCREENSAVER_PLUGIN=NO
  -DBUILD_MARBLE_APPS=NO
  -DBUILD_MARBLE_EXAMPLES=NO
  -DBUILD_MARBLE_TESTS=NO
  -DBUILD_MARBLE_TOOLS=NO
  -DBUILD_TESTING=NO
  -DBUILD_WITH_DBUS=NO
  -DMARBLE_EMPTY_MAPTHEME=YES
  -DMOBILE=NO
  -DWITH_DESIGNER_PLUGIN=NO
  -DWITH_Phonon=NO
  -DWITH_Qt5Location=NO
  -DWITH_Qt5Positioning=NO
  -DWITH_Qt5SerialPort=NO
  -DWITH_ZLIB=NO
  -DWITH_libgps=NO
  -DWITH_libshp=NO
  -DWITH_libwlocate=NO
  -DCMAKE_CXX_COMPILER=**C:\Qt\Tools\mingw730_64\bin\\**
  -DCMAKE_INSTALL_PREFIX=**F:\OneDrive\Dokumente\GitHub\Marble-release\\**
  -DPACKAGE_ROOT_PREFIX=**F:\OneDrive\Dokumente\GitHub\Marble-release\\**
  -DEXEC_INSTALL_PREFIX=**F:\OneDrive\Dokumente\GitHub\Marble-release\\**
  -DCMAKE_PREFIX_PATH=**C:\Qt\5.12.11\mingw73_64\\**
  
  **F:\OneDrive\Dokumente\GitHub\marble\\**

adjust **dir names** to your requirements.

convenience command (using dirs from above):

```
C:\Qt\Tools\CMake_64\bin\cmake  -DCMAKE_BUILD_TYPE=Release  -DSTATIC_BUILD=TRUE  -DQTONLY=TRUE  -DBUILD_MARBLE_EXAMPLES=NO  -DBUILD_INHIBIT_SCREENSAVER_PLUGIN=NO  -DBUILD_MARBLE_APPS=NO  -DBUILD_MARBLE_EXAMPLES=NO  -DBUILD_MARBLE_TESTS=NO  -DBUILD_MARBLE_TOOLS=NO  -DBUILD_TESTING=NO  -DBUILD_WITH_DBUS=NO  -DMARBLE_EMPTY_MAPTHEME=YES  -DMOBILE=NO  -DWITH_DESIGNER_PLUGIN=NO  -DWITH_Phonon=NO  -DWITH_Qt5Location=NO  -DWITH_Qt5Positioning=NO  -DWITH_Qt5SerialPort=NO  -DWITH_ZLIB=NO  -DWITH_libgps=NO  -DWITH_libshp=NO  -DWITH_libwlocate=NO  -DCMAKE_CXX_COMPILER=C:\Qt\Tools\mingw730_64\bin\  -DCMAKE_INSTALL_PREFIX=F:\OneDrive\Dokumente\GitHub\Marble-release\  -DPACKAGE_ROOT_PREFIX=F:\OneDrive\Dokumente\GitHub\Marble-release\  -DEXEC_INSTALL_PREFIX=F:\OneDrive\Dokumente\GitHub\Marble-release\  -DCMAKE_PREFIX_PATH=C:\Qt\5.12.11\mingw73_64\  F:\OneDrive\Dokumente\GitHub\marble\
```

**Reply by albar965** [reply]
The Marble cmake "make install" does not work properly on Windows and macOS. No idea why. You have to copy some files manually on Windows. I have a script for copying and will post this later.

**Reply by u-an-i** [reply]
thank you

**Reply by albar965** [reply]
Sorry have no script. Only for macOS.

Simply copy all headers into the C:\Projekte\marble-release\include\marble|astro dir flat (i.e. no subfolders). Also from build-marble-release\src\lib\astro|marble

Feel free to fix this if you have a clue about cmake. I tried.

**Reply by u-an-i** [reply]
i did. cmake didn't create any lib as far as i found. Marble's INSTALL file states to run `make` und `make install` afterwards from the build directory, the only make i have is C:\Qt\Tools\mingw730_64\bin\mingw32-make.exe (apart from cmake and qmake). Running that results in a message stating "no make target" or similar which is correct as far as I saw: cmake didn't create any makefile.

**Reply by u-an-i** [reply]
and the reason might be although i stated MinGW as the C++ compiler to use, cmake uses msvc! and appears to have created VS project files. I try to find out.

**Reply by u-an-i** [reply]
i now use cmake GUI which comes with Qt

<img width="858" alt="2021-06-13" src="https://user-images.githubusercontent.com/84718885/121792208-48de8980-cbf2-11eb-919f-22521c8b31e9.png">
<img width="858" alt="2021-06-13 (1)" src="https://user-images.githubusercontent.com/84718885/121792209-4bd97a00-cbf2-11eb-93b2-d62ba3dc6c9b.png">

which after selecting only source and target, allows to configure via a wizard after clicking configurre. Choosing MingW makefiles and compiler destination allowed me to set C and C++ compiler of MinGW. Then, every click of configure brought up more configuration options i had to take care of as described in the error message log until marble finally configured.

in the build directory i then entered the path to mingw make and pressed enter and the mingw make install.

I now have libraries inside my release folder. header i already copied to inc.

**Reply by albar965** [reply]
Seems they improved the cmake support in Qt Creator. This was a steaming mess a few versions ago.
Anyway, thanks for feedback and the screenshots.

**Reply by KOKAProduktion** [reply]
Hi guys,

just for completeness sake I'll add my steps here:

Host: `Win10 64-bit, x64`, `Visual Studio 2017`

- Read following instructions:
    - [littlenavmap/wiki/Compiling](https://github.com/albar965/littlenavmap/wiki/Compiling)
    - [littlenavmap/BUILD.txt](https://github.com/albar965/littlenavmap/blob/master/BUILD.txt)
    - [atools/BUILD.txt](https://github.com/albar965/atools/blob/master/BUILD.txt)
    - [littlenavmap/build/win/build_release.cmd](https://github.com/albar965/littlenavmap/blob/master/build/win/build_release.cmd)

- Installed the following using the Qt Installer licensed for (L)GPL development:
    - Qt
        - Qt `5.12.9`
            - MinGW `7.3.0 32-bit`
    - Developer and Designer Tools
        - MinGW `7.3.0 32-bit`

- Checked out atools, littlenavmap and marble into `C:\QtProjects` according to https://github.com/albar965/littlenavmap/discussions/687#discussion-3408464 

- Downloaded latest OpenSSL from https://slproweb.com/products/Win32OpenSSL.html
  - Copied installed OpenSSL content into C:\QtProjects\openssl-1.1.1d-win32-mingw to avoid altering the littlenavmap.pro file 

https://github.com/albar965/littlenavmap/blob/c4efb60b154804677fa43f5c261e5a8811a284d3/littlenavmap.pro#L118

- Ran: (Win Prompt & PowerShell mixed, sry)
```
set PATH=%PATH%;C:\Qt\5.12.9\mingw73_32\bin;C:\Qt\Tools\mingw730_32\bin
cd C:\QtProjects

-------------
# Release

mkdir build-marble-release

cd build-marble-release

cmake -Wno-dev -G "MinGW Makefiles" -DQTONLY=TRUE -DQT5BUILD=TRUE -DWITH_quazip=FALSE -DWITH_libshp=FALSE -DWITH_libgps=FALSE -DWITH_libwlocate=FALSE -DWITH_QtLocation=FALSE -DWITH_liblocation=FALSE -DCMAKE_INSTALL_PREFIX="C:\QtProjects\Marble-release" -DCMAKE_BUILD_TYPE=Release ..\marble

mingw32-make -j2
mingw32-make install

mkdir C:\QtProjects\Marble-release\include

mkdir C:\QtProjects\Marble-release\include\astro

Get-ChildItem -Path "C:\QtProjects\marble\src\lib\astro\*" -Include *.h -Recurse | Copy-Item -Destination C:\QtProjects\Marble-release\include\astro

mkdir C:\QtProjects\Marble-release\include\marble

Get-ChildItem -Path "C:\QtProjects\marble\src\lib\marble\*" -Include *.h -Recurse | Copy-Item -Destination C:\QtProjects\Marble-release\include\marble

copy C:\QtProjects\build-marble-release\src\lib\astro\astrolib_export.h C:\QtProjects\Marble-release\include\astro
copy C:\QtProjects\build-marble-release\src\lib\marble\marble_export.h C:\QtProjects\Marble-release\include\marble

-------------
# Debug

mkdir build-marble-debug

cd build-marble-debug

cmake -Wno-dev -G "MinGW Makefiles" -DQTONLY=TRUE -DQT5BUILD=TRUE -DWITH_quazip=FALSE -DWITH_libshp=FALSE -DWITH_libgps=FALSE -DWITH_libwlocate=FALSE -DWITH_QtLocation=FALSE -DWITH_liblocation=FALSE -DCMAKE_INSTALL_PREFIX="C:\QtProjects\Marble-debug" -DCMAKE_BUILD_TYPE=Debug ..\marble

mingw32-make -j2
mingw32-make install

mkdir C:\QtProjects\Marble-debug\include

mkdir C:\QtProjects\Marble-debug\include\astro

Get-ChildItem -Path "C:\QtProjects\marble\src\lib\astro\*" -Include *.h -Recurse | Copy-Item -Destination C:\QtProjects\Marble-debug\include\astro

mkdir C:\QtProjects\Marble-debug\include\marble

Get-ChildItem -Path "C:\QtProjects\marble\src\lib\marble\*" -Include *.h -Recurse | Copy-Item -Destination C:\QtProjects\Marble-debug\include\marble

copy C:\QtProjects\build-marble-debug\src\lib\astro\astrolib_export.h C:\QtProjects\Marble-debug\include\astro
copy C:\QtProjects\build-marble-debug\src\lib\marble\marble_export.h C:\QtProjects\Marble-debug\include\marble
```
- Imported, qmake'd and compiled atools and littlenavmap inside Qt Creator
    - Set atools as build dependency of littlenavmap
    - Changed default build folder names to the predefined ones:
        - atools
            - build-atools-MinGWXX.XX-Desktop-XX-something-Debug -> **build-atools-debug**
            - build-atools-MinGWXX.XX-Desktop-XX-something-Release-> **build-atools-release**
        - littlenavmap
            - build-littlenavmap-MinGWXX.XX-Desktop-XX-something-Debug -> **build-littlenavmap-debug**
            - build-littlenavmap-MinGWXX.XX-Desktop-XX-something-Release-> **build-littlenavmap-release**

- Adjusted and ran: [/build/win/build_release.cmd](https://github.com/albar965/littlenavmap/blob/c4efb60b154804677fa43f5c261e5a8811a284d3/build/win/build_release.cmd) for the deployment version

I hope this proves helpful to someone! Thanks for the precise instructions @albar965!

Cheers

**Reply by KOKAProduktion** [reply]
Whoops, this might should have gone to: https://github.com/albar965/littlenavmap/discussions/692, sry!

**Reply by u-an-i** [reply]
thank you a lot anyways!
